CRM/Core/Payment/OmnipayMultiProcessor.php

```diff
@@ -639,7 +639,7 @@ public function handlePaymentNotification() {
     if (empty(CRM_Core_Session::singleton()->getLoggedInContactID())) {
       // Don't be too efficient on processing the ipn return.
       // So far the best way of telling the difference is the session.
-      sleep(45);
+      // sleep(45);
```

Comment:
So I have 2 thoughts here - fundamental issue is ipn & browser returning at the same time, starting processing the payment at the same time & sending 2 emails. 
1) We could put ipn_delay_time in the metadata . Some processors are more concurrent than others making it more or less of an issue.
2) we could use a mysql lock - that mostly works prior to mysql 5.7 & can be totally solid in 5.7 with perhaps a small patch - CRM_Core_Lock I believe

Reply:
I don't fully understand the process... But from the sagepay side they expect a response within 1 second otherwise their IPN request fails.  (I spent hours trying to work out why it was timing out on their side, until I found that in their docs somewhere).

Sagepay is a 4-step process as we:
1. Redirect to sagepay servers to make payment.
2. Sagepay notifies CiviCRM of payment status and requests a redirect URL (via IPN).
3. Sagepay receives IPN response (redirectURL) from CiviCRM within 1 second of request (otherwise it retries and times out).
4. Sagepay redirects to redirectURL.

I think most redirect processors are only 3-steps (ie don't have separate steps 2-3) so maybe this is somewhat unique to sagepay?  I don't generally like the idea of arbitrary delays in code (ie. sleep) so if it makes sense to use a mysql lock that might be better?  Though I don't really understand the code and reason for the delay :-)

Reply:
@mattwire I don't suppose you are using mysql 5.7.5+? I feel like using mysql locking is an appropriate way to prevent the 2 processes both processing the ipn at the same time - however, if only works properly with mysql 5.7.5+ 

Some discussion & links here https://github.com/civicrm/civicrm-core/pull/11720 (but I have added a version of this to extension)

Reply:
Thanks.  I've rebased sagepay branch against master, looking much cleaner now.  And I've raised PR #47 to set ipn delay for sagepay